DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,768 @@
+In practice most throw expressions will just look like `throw new Exception()`, but there is a chance that disambiguation is needed when operator overloading is used:
+```D
+throw E0 + E1 => (throw E0) + E1
+throw E0 = E1 => (throw E0) = E1
+```
+
+**(5) A function that returns `auto` may be inferred `noreturn` if every leaf ends in an endless loop or noreturn expression.**
+
+Control flow exiting the function counts as `void`, not `noreturn` since users are used to implicit `return` statements at the end of a `void` function.
+```D
+// return type `void`
+auto main() {
+    import std;
+    writeln("hello world");
+}
+
+// return type `noreturn`
+auto foo(int x) {
+    if (x > 0) {
+        throw new Exception("");
+    } else if (x < 0) {
+        while(true) {}
+    } else {
+        cast(noreturn) main();
+    }
+}
+```
+
+### Properties of `noreturn`:
+
+```D
+noreturn.mangleof == "!";
+```
+This choice is mostly arbitrary (as long as it's unique it's okay for name mangling), but using ! will be familiar to Rust users.
+
+Currently `typeof(null).mangleof` == "n".
+This will be changed to "p!" (pointer to bottom type).
+This might cause linking errors with separate compilation when different translation units are compiled with different D versions.
+However, actually using typeof(null) in a function signature is extremely uncommon outside templates, so in practice this issue won't manifest.
+
+```D
+noreturn.sizeof == 0;
+```
+What is the size of the bottom type? Some considerations are:
+- If something of type A* converts to something of type B* without problems, then one would expect B.sizeof <= A.sizeof. This would imply that noreturn.sizeof >= size_t.max. [(Argued by Timon Gehr)](https://forum.dlang.org/post/okfmt3$vdi$1@digitalmars.com)
+- A boolean requires log2(2)=1 bits storage, a unit type log2(1)=0 bits. Since the bottom type has 0 values, it requires log2(0) bits storage, which is undefined. It approaches -Infinity in the limit.
+- There is no meaningful size, so the size is the bottom value `⊥` which implicitly converts to a `size_t`. Any time `noreturn.sizeof` is used, the program crashes.
+- A `union` has the size of the largest member. Adding a `noreturn` field to a `union` never increases the size, so its size is 0.
+
+The last definition seems the simplest and most useful, so it is the proposed one.
+
+```D
+noreturn.alignof == 0;
+```
+Every pointer to type `T` with alignof `n` must have an address `n*k` for some integer `k`.
+Choosing `n = 0` forces every `T*` to be at address 0, which is the case for `typeof(null)`.
+
+```D
+noreturn.init == assert(0);
+```
+The type `noreturn` has no values, so there is no `init` value either (or the init value is the bottom value `⊥`).
+Any time `noreturn.init` appears in code, it gets lowered to `assert(0)`.
+
+```D
+noreturn*.sizeof == 0
+```
+Currently `typeof(null).sizeof` equals `size_t.sizeof`, but since it's a unit type it doesn't need any storage.
+
```

Comment:
It doesn't _need_ any storage, but `struct Unit{ }` which is also an unit type still takes one `byte`, so this rationale is insufficient.